src/ngx_http_lua_pipe.c

```diff
@@ -0,0 +1,2473 @@
+
+    pipe_ctx->buf_in = cl;
+    b = pipe_ctx->buf_in->buf;
+    b->last = ngx_copy(b->last, data, len);
+
+    rc = ngx_http_lua_pipe_write(pipe, pipe_ctx);
+    if (rc == NGX_ERROR) {
+        goto error;
+    }
+
+    if (rc == NGX_OK) {
+        return len;
+    }
+
+    /* rc == NGX_AGAIN */
+    wait_co_ctx = ctx->cur_co_ctx;
+    pipe_ctx->c->data = wait_co_ctx;
+
+    wev->handler = ngx_http_lua_pipe_resume_write_handler;
+    if (ngx_handle_write_event(wev, 0) != NGX_OK) {
+        pipe_ctx->err_type = PIPE_ERR_ADD_WRITE_EV;
+        goto error;
+    }
+
+    wait_co_ctx->data = proc;
+    wait_co_ctx->cleanup = ngx_http_lua_pipe_proc_write_cleanup;
+    timeout = proc->write_timeout;
+
+    if (timeout > 0) {
+        ngx_add_timer(wev, timeout);
+        ngx_log_debug5(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                       "lua pipe add timer for writing: %d(ms) process:%p "
+                       "pid:%P pipe:%p ev:%p", timeout, proc, proc->_pid, pipe,
+                       wev);
+    }
+
+    ngx_log_debug3(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua pipe write yielding process:%p pid:%P pipe:%p", proc,
+                   proc->_pid, pipe);
+
+    return NGX_AGAIN;
+
+error:
+
+    ngx_http_pipe_put_error(pipe_ctx, errbuf, errbuf_size);
+    return NGX_ERROR;
+}
+
+
+/*
+ * ngx_http_lua_ffi_pipe_get_write_result should only be called just after
+ * ngx_http_lua_ffi_pipe_proc_write, so we omit most of the sanity check already
+ * done in ngx_http_lua_ffi_pipe_proc_write.
+ */
+ssize_t ngx_http_lua_ffi_pipe_get_write_result(ngx_http_request_t *r,
+    ngx_http_lua_ffi_pipe_proc_t *proc, u_char *errbuf, size_t *errbuf_size)
+{
+    ngx_http_lua_pipe_t                *pipe;
+    ngx_http_lua_pipe_ctx_t            *pipe_ctx;
+
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua pipe get write result process:%p pid:%P", proc,
+                   proc->_pid);
+
+    pipe = proc->pipe;
+    pipe_ctx = pipe->stdin_ctx;
+
+    if (pipe_ctx->err_type) {
+        ngx_http_pipe_put_error(pipe_ctx, errbuf, errbuf_size);
+        return NGX_ERROR;
+    }
+
+    return pipe_ctx->rest;
+}
+
+
+int
+ngx_http_lua_ffi_pipe_proc_wait(ngx_http_request_t *r,
+    ngx_http_lua_ffi_pipe_proc_t *proc, char **reason, int *status,
+    u_char *errbuf, size_t *errbuf_size)
+{
+    int                                 rc;
+    ngx_rbtree_node_t                  *node;
+    ngx_http_lua_ctx_t                 *ctx;
+    ngx_http_lua_pipe_t                *pipe;
+    ngx_http_lua_co_ctx_t              *wait_co_ctx;
+    ngx_http_lua_pipe_node_t           *pipe_node;
+
+    rc = ngx_http_lua_pipe_get_lua_ctx(r, &ctx, errbuf, errbuf_size);
+    if (rc != NGX_OK) {
+        return rc;
+    }
+
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua pipe wait process:%p pid:%P", proc, proc->_pid);
+
+    pipe = proc->pipe;
+    if (pipe == NULL || pipe->closed) {
```

Comment:
I think we should check for `pipe->dead` here instead?

Reply:
@thibaultcha 
We may wait for a `dead` process (the process is dead just before we call `wait`).

Reply:
So we may wait for a `dead` process, but we may not wait when the pipe is `closed`? Why so?

Reply:
The pipe is `closed` when it is manually 'closed' by the users (via `proc:kill` or `proc:wait`).

Reply:
Only `proc:wait` or `__gc` seem to set `closed`. Nonetheless, I see the reason :+1: